File: packages/create-crucible-agent/templates/ts/strategy.ts
// strategy.ts — your trading brain.
//
// Edit `prompt.md` to change the system prompt without touching code.
// Edit `decide()` to swap the model, add tools, change parsing, etc.
// Set LLM_PROVIDER and LLM_MODEL in crucible.env to pick any provider —
// no code change needed for simple model swaps.

import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import { generateText, type LanguageModel } from "ai";

export type AgentDecision = { kind: string; qty: bigint; reasoning: string };

export const meta = {
  model:        process.env["LLM_MODEL"]    ?? "claude-haiku-4-5",
  framework:    process.env["LLM_FRAMEWORK"] ?? "vercel-ai-sdk",
  agentVersion: process.env["AGENT_VERSION"] ?? "0.1.0",
};

const SYSTEM_PROMPT = readFileSync(resolve(process.cwd(), "prompt.md"), "utf8");

// Lazy-resolve the provider so users only need to install the SDK they use.
async function resolveModel(): Promise<LanguageModel> {
  const provider = (process.env["LLM_PROVIDER"] ?? "anthropic").toLowerCase();
  const model    = process.env["LLM_MODEL"] ?? "claude-haiku-4-5";

  switch (provider) {
    case "anthropic": {
      const { anthropic } = await import("@ai-sdk/anthropic");
      return anthropic(model);
    }
    case "openai": {
      const { openai } = await import("@ai-sdk/openai");
      return openai(model);
    }
    case "google": {
      const { google } = await import("@ai-sdk/google");
      return google(model);
    }
    case "mistral": {
      const { mistral } = await import("@ai-sdk/mistral");
      return mistral(model);
    }
    case "openrouter":
    case "openai-compatible":
    case "ollama": {
      const { createOpenAICompatible } = await import("@ai-sdk/openai-compatible");
      const baseURL = process.env["LLM_BASE_URL"] ??
        (provider === "openrouter" ? "https://openrouter.ai/api/v1"
         : provider === "ollama"    ? "http://localhost:11434/v1"
         : (() => { throw new Error("LLM_BASE_URL required for openai-compatible"); })());
      const apiKey = process.env["LLM_API_KEY"] ?? process.env["OPENROUTER_API_KEY"] ?? "not-needed";
      const client = createOpenAICompatible({ name: provider, baseURL, apiKey });
      return client(model);
    }
    default:
      throw new Error(`Unknown LLM_PROVIDER "${provider}". Supported: anthropic, openai, google, mistral, openrouter, ollama, openai-compatible.`);
  }
}

let _model: LanguageModel | undefined;
async function model(): Promise<LanguageModel> {
  if (!_model) _model = await resolveModel();
  return _model;
}

export async function decide(observation: Record<string, unknown>): Promise<AgentDecision> {
  const { text } = await generateText({
    model: await model(),
    system: SYSTEM_PROMPT,
    prompt: JSON.stringify(observation),
    maxOutputTokens: 256,
  });

  const cleaned = text.replace(/^```(?:json)?\s*|\s*```$/gm, "").trim();
  const j = JSON.parse(cleaned) as { kind: string; qty?: string; reasoning?: string };
  return { kind: j.kind, qty: BigInt(j.qty ?? "0"), reasoning: j.reasoning ?? "" };
}

File: packages/create-crucible-agent/templates/py/strategy.py
"""strategy.py — your trading brain.

Edit `prompt.md` to change the system prompt without touching code.
Edit `decide()` to swap the model, add tools, change parsing, etc.
Set LLM_PROVIDER / LLM_MODEL in crucible.env to pick any provider —
no code change needed for simple model swaps.

Uses `litellm` for unified provider access (anthropic, openai, gemini, mistral,
openrouter, ollama, bedrock, …) — see https://docs.litellm.ai/docs/providers.
"""
import json
import os
from pathlib import Path
from typing import Any

import litellm

PROMPT = Path(__file__).parent.joinpath("prompt.md").read_text(encoding="utf-8")

# Provider prefix in front of the model id is litellm's routing convention.
# anthropic/claude-haiku-4-5, openai/gpt-4o-mini, gemini/gemini-2.0-flash, etc.
PROVIDER = os.environ.get("LLM_PROVIDER", "anthropic")
MODEL_ID = os.environ.get("LLM_MODEL", "claude-haiku-4-5")
ROUTED_MODEL = MODEL_ID if "/" in MODEL_ID else f"{PROVIDER}/{MODEL_ID}"

# openai-compatible / ollama / openrouter: pass base_url through.
EXTRA_KWARGS: dict[str, Any] = {}
if base := os.environ.get("LLM_BASE_URL"):
    EXTRA_KWARGS["api_base"] = base
if key := os.environ.get("LLM_API_KEY"):
    EXTRA_KWARGS["api_key"] = key

META = {
    "model": MODEL_ID,
    "framework": os.environ.get("LLM_FRAMEWORK", "litellm"),
    "agentVersion": os.environ.get("AGENT_VERSION", "0.1.0"),
}


def decide(obs: dict[str, Any]) -> tuple[str, int, str]:
    """Receive an observation; return (kind, qty_in_wei, reasoning)."""
    r = litellm.completion(
        model=ROUTED_MODEL,
        max_tokens=256,
        messages=[
            {"role": "system", "content": PROMPT},
            {"role": "user", "content": json.dumps(obs)},
        ],
        **EXTRA_KWARGS,
    )
    txt = r["choices"][0]["message"]["content"].strip()
    if txt.startswith("```"):
        txt = txt.strip("`").lstrip("json").strip()
    j = json.loads(txt)
    return j["kind"], int(j.get("qty", "0")), j.get("reasoning", "")
